File: 110-BalancedBinaryTree/110-BalancedBinaryTree.py
# Last updated: 2/8/2026, 12:13:37 PM
1class Solution:
2  def isBalanced(self, root: TreeNode | None) -> bool:
3    if not root:
4      return True
5
6    def maxDepth(root: TreeNode | None) -> int:
7      if not root:
8        return 0
9      return 1 + max(maxDepth(root.left), maxDepth(root.right))
10
11    return (abs(maxDepth(root.left) - maxDepth(root.right)) <= 1 and
12            self.isBalanced(root.left) and
13            self.isBalanced(root.right))

File: 110-BalancedBinaryTree/110-BalancedBinaryTree.cpp
// Last updated: 6/2/2025, 6:39:59 PM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        function<int(TreeNode*)> height = [&](TreeNode* root) {
            if (!root) {
                return 0;
            }
            int l = height(root->left);
            int r = height(root->right);
            if (l == -1 || r == -1 || abs(l - r) > 1) {
                return -1;
            }
            return 1 + max(l, r);
        };
        return height(root) >= 0;
    }
};